File: ts/src/calibrator.ts
/**
 * TokenCalibrator (writer) — the stateful producer side. Wraps ONE model's
 * data-only accumulator: feed it real usage via `observe`, read back solved
 * `rates()` or the raw `toMatrix()` for persistence. All the math is the pure
 * `calibration.ts` functions; this is just per-instance state + convenience.
 */
import { estimateTokens, type TokenRates } from './buckets.js';
import {
  accumulate,
  emptyAccumulator,
  isValidAccumulator,
  ratesFromAccumulator,
  type AccumulateOptions,
  type RatesOptions,
  type TokenAccumulator,
} from './calibration.js';

export type TokenCalibratorOptions = RatesOptions & AccumulateOptions & {
  /** Resume from a persisted matrix (e.g. `toMatrix()` from a prior run).
   * A corrupt matrix is ignored and a fresh accumulator is used.
   */
  accumulator?: TokenAccumulator;
};

/**
 * Stateful writer around ONE model's accumulator.
 *
 * const cal = new TokenCalibrator();
 * cal.observe(requestText, realInputTokens);
 * const rates = cal.rates(); // for estimation
 * const matrix = cal.toMatrix(); // { [model]: matrix } → persist
 */
export class TokenCalibrator {
  private acc: TokenAccumulator;
  private readonly forgetting?: number;
  private readonly priorStrength?: number;
  private readonly prior?: TokenRates;

  constructor(opts: TokenCalibratorOptions = {}) {
    this.forgetting = opts.forgetting;
    this.priorStrength = opts.priorStrength;
    this.prior = opts.prior;

    this.acc = opts.accumulator && isValidAccumulator(opts.accumulator)
        ? {
          a: opts.accumulator.a.map((row) => row.slice()),
          g: opts.accumulator.g.slice(),
        }
        : emptyAccumulator();
  }

  /** Fold in one observed round. */
  observe(input: string, realTokens: number): void {
    this.acc = accumulate(this.acc, input, realTokens, {
      forgetting: this.forgetting,
    });
  }

  /** Current solved per-bucket rates. */
  rates(): TokenRates {
    return ratesFromAccumulator(this.acc, {
      priorStrength: this.priorStrength,
      prior: this.prior,
    });
  }

  /** Convenience: estimate directly from this calibrator's current rates.
   * For the per-model path prefer `TokenEstimator`; this is sugar for when
   * you already hold a single calibrator.
   */
  estimate(input: string): number {
    return estimateTokens(input, this.rates());
  }

  /** The raw data-only accumulator, for persistence. Store under a model
   * key: `{ [model]: cal.toMatrix() }`; merge is a plain object spread.
   */
  toMatrix(): TokenAccumulator {
    return {
      a: this.acc.a.map((row) => row.slice()),
      g: this.acc.g.slice(),
    };
  }
}

File: python/src/token_calibrator/calibrator.py
"""
TokenCalibrator (writer) — the stateful producer side. Wraps ONE model's
data-only accumulator: feed it real usage via `observe`, read back solved
`rates()` or the raw `to_matrix()` for persistence. All the math is the pure
`calibration.py` functions; this is just per-instance state + convenience.
"""

from __future__ import annotations

from typing import Dict, List, Optional

from .buckets import estimate_tokens, TokenRates
from .calibration import (
    accumulate,
    empty_accumulator,
    is_valid_accumulator,
    rates_from_accumulator,
    TokenAccumulator,
    AccumulateOptions,
    RatesOptions,
)


class TokenCalibratorOptions(RatesOptions, AccumulateOptions):
    accumulator: Optional[TokenAccumulator] = None


class TokenCalibrator:
    """
    Stateful writer around ONE model's accumulator.

    Usage:
        cal = TokenCalibrator()
        cal.observe(request_text, real_input_tokens)
        rates = cal.rates()  # for estimation
        matrix = cal.to_matrix()  # for persistence
    """

    def __init__(self, opts: Optional[TokenCalibratorOptions] = None) -> None:
        if opts is None:
            opts = {}
        self._forgetting: Optional[float] = opts.get("forgetting")
        self._prior_strength: Optional[float] = opts.get("priorStrength")
        self._prior: Optional[TokenRates] = opts.get("prior")

        acc = opts.get("accumulator")
        if acc is not None and is_valid_accumulator(acc):
            self._acc: TokenAccumulator = {
                "a": [row[:] for row in acc["a"]],
                "g": list(acc["g"]),
            }
        else:
            self._acc = empty_accumulator()

    def observe(self, input_str: str, real_tokens: float) -> None:
        """Fold in one observed round."""
        opts_accum: AccumulateOptions = {}
        if self._forgetting is not None:
            opts_accum["forgetting"] = self._forgetting
        self._acc = accumulate(self._acc, input_str, real_tokens, opts_accum)

    def rates(self) -> TokenRates:
        """Current solved per-bucket rates."""
        opts_rates: RatesOptions = {}
        if self._prior_strength is not None:
            opts_rates["priorStrength"] = self._prior_strength
        if self._prior is not None:
            opts_rates["prior"] = self._prior
        return rates_from_accumulator(self._acc, opts_rates)

    def estimate(self, input_str: str) -> int:
        """Convenience: estimate directly from this calibrator's current rates."""
        return estimate_tokens(input_str, self.rates())

    def to_matrix(self) -> TokenAccumulator:
        """The raw data-only accumulator, for persistence."""
        return {
            "a": [row[:] for row in self._acc["a"]],
            "g": list(self._acc["g"]),
        }
